### PR TITLE
Results view for teacher

### PR DESCRIPTION
**Pull request by makitzei**
Co-authored-by: @jonthus 

Teacher can view assignment performances on a separate page. Some role restrictions (same restrictions as in detailed student reports view) added too.

Only faults now in codecov coverage.

### CODE REVIEW

File: locallib.php [hunk 1]
```diff
@@ -518,6 +519,69 @@ function get_attempt($instanceid) {
...
+        $score = new html_table_cell($attempt->gop_score);
+    }
+    $time = new html_table_cell('TODO');
+    $tries = new html_table_cell('TODO');
+
+    $urltext = page_url(3, $id, $d, $attempt->userid);
+    $urllink = html_writer::link($urltext, get_string('results_link', 'digitala'));
+    $reportlink = new html_table_cell($urllink);
+
+    $cell->attributes['class'] = 'showFine';
```
Comment (FinThunderstorm) [inline locallib.php:578]: Where is this variable `cell` defined? This is causing the bug shown in the image. It does function with commenting this row out.

File: locallib.php [hunk 2]
```diff
@@ -518,6 +519,69 @@ function get_attempt($instanceid) {
...
+    $type = new html_table_cell($digitalatype->attempttype);
+    if ($attempt->holistic) {
+        $score = new html_table_cell($attempt->holistic);
+    } else {
+        $score = new html_table_cell($attempt->gop_score);
+    }
+    $time = new html_table_cell('TODO');
+    $tries = new html_table_cell('TODO');
+
+    $urltext = page_url(3, $id, $d, $attempt->userid);
```
Comment (FinThunderstorm) [inline locallib.php:574]: could this be something like `report.php?mode=detail&student=1`? I will discuss about this with Topi tomorrow morning, when we are taking a deeper look into showing detailed feedback.

**Review by FinThunderstorm — CHANGES_REQUESTED** [review]
image
Some weird bugs are appearing.

About link policy, could all renderings for teachers ability to check answers to be extracted from view.php and made into own file `report.php`? Like in [Quiz-activity?](https://github.com/moodle/moodle/blob/master/mod/quiz/report.php)

File: locallib.php (hunk 1 above)
Comment (makitzei) [reply locallib.php:578]: Oh yeah, this one is left there by mistake. And it turns out it was unnecessary too.

File: locallib.php [hunk 3]
```diff
@@ -31,9 +31,10 @@
  * @param number $page number of the step
  * @param number $id id of the course module
  * @param number $d id of the activity instance
+ * @param number $studentid id for the student to render report for teacher
  */
-function page_url($page, $id, $d) {
-    return new moodle_url('/mod/digitala/view.php', array('id' => $id, 'd' => $d, 'page' => $page));
+function page_url($page, $id, $d, $studentid=null) {
+    return new moodle_url('/mod/digitala/view.php', array('id' => $id, 'd' => $d, 'page' => $page, 'studentid' => $studentid));
```
[the remark this answers is not in the document]
Comment (makitzei) [reply locallib.php]: Changed this to have mode values 'detail' and 'overview' as suggested elsewhere.

File: locallib.php (hunk 2 above)
Comment (makitzei) [reply locallib.php]: Good suggestion, done so.

File: renderer.php [hunk 4]
```diff
@@ -49,6 +51,14 @@ protected function render_digitala_progress_bar(digitala_progress_bar $progressb
         $out .= create_progress_bar_spacer($spacers['right']);
         $out .= create_progress_bar_step('report', 2, $progressbar->id, $progressbar->d, $progressbar->currpage);
         $out .= end_progress_bar();
+        // Temporary link to student results page.
+
+        $systemcontext = context_system::instance();
+        // First version of user permissions check.
+        if (has_capability('mod/digitala:addinstance', $systemcontext)) {
```
[the remark this answers is not in the document]
Comment (jonthus) [reply renderer.php:58]: This was done by makitzei already in another location, but this line of code doesn't exist anymore.

File: locallib.php [hunk 5]
```diff
@@ -533,24 +576,126 @@ function save_attempt($assignment, $filename, $evaluation, $recordinglength) {
...
+function create_result_row($attempt, $instanceid, $id, $d) {
+    $user = get_user($attempt->userid);
+
+    $username = $user->firstname . ' ' . $user->lastname;
+    if ($attempt->holistic) {
+        $score = $attempt->holistic;
+    } else {
+        $score = $attempt->gop_score;
+    }
+    $time = ($attempt->recordinglength . ' s');
```
Comment (FinThunderstorm) [inline locallib.php:689]:
```suggestion
    $time = get_($attempt->recordinglength);
```
Does converting this to format `02:27` (2 minutes and 27 seconds) make more sense than showing `147 s`?

**Review by FinThunderstorm — COMMENTED** [review]
Leaved one idea that came up in mind when looking up this

File: locallib.php (hunk 5 above)
Comment (jonthus) [reply locallib.php:689]: It does look a lot better, will commit this suggestion.

**Pull request merged**